src/reporting/clients/ODataClient.ts

```diff
@@ -0,0 +1,86 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+import { AccessToken } from "@itwin/core-bentley";
+import { RequiredError } from "../interfaces/Errors";
+import { PagedResponseLinks } from "../interfaces/Links";
+import { ODataResponse, ODataItem, ODataEntityResponse, ODataEntityValue } from "../interfaces/OData";
+import { EntityListIterator } from "../iterators/EntityListIterator";
+import { EntityListIteratorImpl } from "../iterators/EntityListIteratorImpl";
+import { getEntityCollectionPage, Collection } from "../iterators/IteratorUtil";
+import { OperationsBase } from "../OperationsBase";
+import { IOdataClient } from "./IODataClient";
+
+export class ODataClient extends OperationsBase implements IOdataClient{
+  public async getODataReport(accessToken: AccessToken, reportId: string): Promise<ODataResponse> {
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return this.fetchData<ODataResponse>(url, requestOptions);
+  }
+
+  public async getODataReportEntityPage(accessToken: AccessToken, reportId: string, odataItem: ODataItem, sequence: number) : Promise<ODataEntityResponse> {
+    const segments = odataItem.url.split('/');  // region, manifestId, entityType
+    if (segments.length !== 3) {
+      throw new RequiredError(
+        'odataItem',
+        'odata item was invalid when calling updateCalculatedProperty.',
+      );
+    }
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}/${odataItem.url}?sequence=${encodeURIComponent(sequence)}`;
+    const requestOptions: RequestInit = this.createRequest("GET", accessToken);
+    return this.fetchData(url, requestOptions);
+  }
+
+  public async getODataReportEntities(accessToken: AccessToken, reportId: string, odataItem: ODataItem): Promise<Array<ODataEntityValue>> {
+    const segments = odataItem.url.split('/');  // region, manifestId, entityType
+    if (segments.length !== 3) {
+      throw new RequiredError(
+        'odataItem',
+        'odata item was invalid when calling updateCalculatedProperty.',
+      );
+    }
+    const reportData: Array<ODataEntityValue> = [];
+    const oDataReportEntitiesIt = this.getODataReportEntitiesIterator(accessToken, reportId, odataItem);
+    for await(const oDataReportEntity of oDataReportEntitiesIt) {
+      reportData.push(oDataReportEntity);
+    }
+    return reportData;
+  }
+
+  public getODataReportEntitiesIterator(accessToken: AccessToken, reportId: string, odataItem: ODataItem): EntityListIterator<ODataEntityValue> {
+    const url = `${this.basePath}/odata/${encodeURIComponent(reportId)}/${odataItem.url}`;
```

Comment:
In the methods above `odataItem.url` is validated. Validate here as well.

Reply:
It is now validated. Since getODataReportEntitiesIterator is not an async function an arrow function was used.